x-pack/elastic-agent/pkg/agent/application/upgrade/rollback.go

```diff
@@ -0,0 +1,158 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/elastic/beats/v7/libbeat/common/backoff"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/application/paths"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/control"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/control/client"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/logger"
+)
+
+const (
+	watcherSubcommand  = "watch"
+	maxRestartCount    = 5
+	restartBackoffInit = 5 * time.Second
+	restartBackoffMax  = 90 * time.Second
+)
+
+// Rollback rollbacks to previous version which was functioning before upgrade.
+func Rollback(ctx context.Context, prevHash, currentHash string) error {
+	// change symlink
+	if err := ChangeSymlink(ctx, prevHash); err != nil {
+		return err
+	}
+
+	// revert active commit
+	if err := UpdateActiveCommit(prevHash); err != nil {
+		return err
+	}
+
+	// Restart
+	if err := restartAgent(ctx); err != nil {
+		return err
+	}
+
+	// cleanup everything except version we're rolling back into
+	return Cleanup(prevHash, true)
+}
+
+// Cleanup removes all artifacts and files related to a specified version.
+func Cleanup(currentHash string, removeMarker bool) error {
+	<-time.After(afterRestartDelay)
+
+	// remove upgrade marker
+	if removeMarker {
+		if err := CleanMarker(); err != nil {
+			return err
+		}
+	}
+
+	// remove data/elastic-agent-{hash}
+	dataDir, err := os.Open(paths.Data())
+	if err != nil {
+		return err
+	}
+
+	subdirs, err := dataDir.Readdirnames(0)
+	if err != nil {
+		return err
+	}
+
+	dirPrefix := fmt.Sprintf("%s-", agentName)
+	currentDir := fmt.Sprintf("%s-%s", agentName, currentHash)
+	for _, dir := range subdirs {
+		if dir == currentDir {
+			continue
+		}
+
+		if !strings.HasPrefix(dir, dirPrefix) {
+			continue
+		}
+
+		hashedDir := filepath.Join(paths.Data(), dir)
+		if cleanupErr := os.RemoveAll(hashedDir); cleanupErr != nil && !isErrorExpected(cleanupErr) {
+			err = multierror.Append(err, cleanupErr)
+		}
+	}
+
+	return err
+}
+
+func isErrorExpected(err error) bool {
+	// cannot remove self, this is expected
+	// fails with  remove {path}}\elastic-agent.exe: Access is denied
+	if runtime.GOOS == "windows" && strings.Contains(err.Error(), "elastic-agent.exe") && strings.Contains(err.Error(), "Access is denied") {
+		return true
```

Comment:
I have code in the Uninstall command, that tries to handle this. It might be good to generalize that so it can be used here.

It uses a spawned `cmd.exe` to cleanup the `elastic-agent.exe` after uninstall. Same could be used here by the watcher.